dev/tools/controllerbuilder/pkg/codegen/typegenerator.go

```diff
@@ -41,45 +41,48 @@ var protoMessagesNotMappedToGoStruct = map[string]string{
 
 type TypeGenerator struct {
 	generatorBase
-	goPackage             string
-	resourceProtoFullName string
-	visitedMessages       []protoreflect.MessageDescriptor
+	api             *protoapi.Proto
+	goPackage       string
+	visitedMessages []protoreflect.MessageDescriptor
 }
 
-func NewTypeGenerator(goPackage string, outputBaseDir, resourceProtoFullName string) *TypeGenerator {
+func NewTypeGenerator(goPackage string, outputBaseDir string, api *protoapi.Proto) *TypeGenerator {
 	g := &TypeGenerator{
-		goPackage:             goPackage,
-		resourceProtoFullName: resourceProtoFullName,
+		goPackage: goPackage,
+		api:       api,
 	}
 	g.generatorBase.init(outputBaseDir)
 	return g
 }
 
-func (g *TypeGenerator) VisitProto(api *protoapi.Proto) error {
-	if err := g.visitMessages(api.Files()); err != nil {
+func (g *TypeGenerator) VisitProto(resourceProtoFullName string) error {
+
+	descriptor, err := g.api.Files().FindDescriptorByName(protoreflect.FullName(resourceProtoFullName))
+	if err != nil {
+		return fmt.Errorf("failed to find the proto message %s: %w", resourceProtoFullName, err)
+	}
+	messageDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		return fmt.Errorf("unexpected descriptor type: %T", descriptor)
+	}
+
+	if err := g.visitMessage(messageDescriptor); err != nil {
 		return err
 	}
 
-	g.writeVisitedMessages()
 	return nil
 }
 
-func (g *TypeGenerator) visitMessages(files *protoregistry.Files) error {
-	messageDesc, err := files.FindDescriptorByName(protoreflect.FullName(g.resourceProtoFullName))
-	if err != nil {
-		return fmt.Errorf("failed to find the proto message %s: %w", g.resourceProtoFullName, err)
-	}
-	msgDesc, ok := messageDesc.(protoreflect.MessageDescriptor)
-	if !ok {
-		return fmt.Errorf("unexpected descriptor type: %T", msgDesc)
-	}
-	//klog.Infof("found message %q", msgDesc.FullName())
+func (g *TypeGenerator) visitMessage(messageDescriptor protoreflect.MessageDescriptor) error {
+	//klog.Infof("found message %q", messageDescriptor.FullName())
+
+	g.visitedMessages = append(g.visitedMessages, messageDescriptor)
 
-	msgs, err := findDependenciesForMessage(msgDesc)
+	msgs, err := findDependenciesForMessage(messageDescriptor)
 	if err != nil {
 		return err
 	}
-	g.visitedMessages = append(msgs, msgDesc)
+	g.visitedMessages = append(g.visitedMessages, msgs...)
```

Comment:
Do we need to sort and remove duplicate after calling this function for multiple resources?

Reply:
Good call.  This list is only used when we write the messages, so I just sort and dedup there now.